packages/vaex-core/vaex/dataframe_protocol.py

```diff
@@ -0,0 +1,489 @@
+        return ordered, is_dictionary, mapping
+    
+    @property
+    def describe_null(self) -> Tuple[int, Any]:
+        """
+        Return the missing value (or "null") representation the column dtype
+        uses, as a tuple ``(kind, value)``.
+        Kind:
+            - 0 : non-nullable
+            - 1 : NaN/NaT
+            - 2 : sentinel value
+            - 3 : bit mask
+            - 4 : byte mask
+        Value : if kind is "sentinel value", the actual value. None otherwise.
+        """
+        _k = _DtypeKind
+        kind = self.dtype[0]
+        value = None
+        if kind == _k.FLOAT:
+            null = 1  # np.nan
+        elif kind == _k.DATETIME:
+            null = 1  # np.datetime64('NaT')
+        elif kind in (_k.INT, _k.UINT, _k.BOOL):
+            # TODO: check if extension dtypes are used once support for them is
+            #       implemented in this procotol code
+            null = 0  # integer and boolean dtypes are non-nullable
+        elif kind == _k.CATEGORICAL:
+            # Null values for categoricals are stored as `-1` sentinel values
+            # in the category date (e.g., `col.values.codes` is int8 np.ndarray)
+            null = 2
+            value = -1
+        else:
+            raise NotImplementedError(f'Data type {self.dtype} not yet supported')
+
+        return null, value
+
+    @property
+    def null_count(self) -> int:
+        """
+        Number of null elements. Should always be known.
+        """
+        return {}
+
+    def num_chunks(self) -> int:
+        """
+        Return the number of chunks the column consists of.
+        """
+        return 1
+
+    def get_chunks(self, n_chunks : Optional[int] = None) -> Iterable['_PandasColumn']:
+        """
+        Return an iterator yielding the chunks.
+        See `DataFrame.get_chunks` for details on ``n_chunks``.
+        """
+        return {}
+    
+    @property
+    def metadata(self, dict_is_cat) -> Dict[str, Any]:
+        """
+        Store specific metadata of the column.
+        """
+        # Boolean if column is categorical or not
+        return dict_is_cat
+    
+    def get_data_buffer(self) -> Tuple[_VaexBuffer, Any]:  # Any is for self.dtype tuple
+        """
+        Return the buffer containing the data.
+        """
+        _k = _DtypeKind
+        if self.dtype[0] in (_k.INT, _k.UINT, _k.FLOAT, _k.BOOL):
+            buffer = _VaexBuffer(self._col.to_numpy())
+            dtype = self.dtype
+        elif self.dtype[0] == _k.CATEGORICAL:
+            codes = self._col.values
+            # If values are subset of labels and are not codes, codes mst be generated
+            if self._col.values[0] in self.labels:
+                for i in self._col.values:
+                    codes[np.where(codes==i)] = np.where(self.labels == i) # if values are same as labels
+            else: 
+                codes = self._col.values # values are already codes for the labels

```

Comment:
I also think this will not be needed when I will be able to use `expressions.codes.`

But for now, if I take this example:
```python
df = vaex.from_arrays(year=[2012, 2015, 2019], weekday=[0, 4, 6])
df = df.categorize('year', min_value=2012, max_value=2019)
df = df.categorize('weekday', labels=['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'])
```
and use `.values` function on `year` and `weekday` column I get two different types of output. In the first case it is a list of labels and in the other case I get the codes.
```
array([2012, 2015, 2019])
array([0, 4, 6])
```
To get the codes for the `year` column I calculate them as in line 402-404.

For now I can't seem to find another way.